Remove Usage Of Deprecated `lines` Method

isomarcte opened the pull request:
Another instance of the deprecated `lines` method was recently added to the code base. This method causes metals to be not build on JDK >= 11.

This commit also updates the CI to build with JDKs 8 and 11 using AdpotOpenJDK. The majority of tests are JDK 11 as default, with a single test using JDK8 as a legacy compatibility smoke test. This should help prevent future additions of this method.

Finally, this commit drops cross building with deprecated scala versions, as they are unable to parse class files generated with bytecode for JDKs > 8.

See: ce9c890c27efec49fd8ad78746e2f406bfd89d3b and 4d78f9cf5a2e6c49d8a6e2474a49a8bbb99e216c for more context.

isomarcte commented:
The failures in CI seem to be timeout issues unrelated to these changes.

olafurpg commented on .github/workflows/ci.yml:
```diff
@@ -12,9 +12,15 @@ jobs:
+        java-version:
+          - "adopt@1.11"
+          - "adopt@1.8"
+          - "openjdk@1.11"
```
What’s the motivation to have two JDK 11 entries?

olafurpg replied:
We should include the jdk name in the job name to make it easier to distinguish in the web ui.

isomarcte replied:
@olafurpg I'm fine with just one. I added two because I had noticed you were using AdoptOpenJDK for the 1.8 builds. I thought it may be wise to use OpenJDK's "official" implementation when available, which it is for 11, and I also wanted to keep parity with what you were already using, hence AdoptOpenJDK. 

That being said, using just one of the two is probably fine.

olafurpg commented on tests/unit/src/test/scala/tests/TreeViewLspSuite.scala:
```diff
@@ -1,9 +1,50 @@
+   * @note this value changes depending on the JVM version in use as some JAR
+   *       files have moved to become modules on JVM > 8.
+   */
+  val expectedLibraries: SortedSet[String] = {
```
Thank you for fixing this! 👏

The biggest reason for delaying the upgrade to jdk 11 in our ci was that tests needed to be updated.

gabro commented:
Thank you @isomarcte for working on this!

@olafurpg what about testing on JDK 11 by the default, and only keeping one “canary” job on jdk 8? My concern is not to slow down the CI too much

olafurpg commented:
@gabro I'm in favor of that approach 👍

gabro commented on .github/workflows/ci.yml:
```diff
@@ -5,16 +5,38 @@ on:
   pull_request:
 jobs:
+  ubuntuUnitWithJDK8And11:
+    name: ubuntu-latest tests
```
can you include the jdk version in the name?

```suggestion
    name: ubuntu-latest tests (${{  matrix.java-version }})
```

or something like this

isomarcte replied:
yep

isomarcte commented:
@gabro @olafurpg this should be good to go to go now. I've changed it to the ubuntu tests with JDK8 and JDK11, while the windows and macOS tests only use JDK11. 

I've also changed it to only use `"openjdk1.11"` so we'll only have one JDK11.

Finally, I switched to use the system property `"java.vm.specification.version"` as I did some reading around and that seems like the most technically correct value to use.

isomarcte commented:
@gabro can you take a look again. I update the other tests to use JDK 11 as well since we want that to be the global default.

isomarcte commented on .github/workflows/ci.yml:
```diff
@@ -57,6 +85,8 @@ jobs:
     steps:
       - uses: actions/checkout@v1
       - uses: olafurpg/setup-scala@v5
```
@gabro @olafurpg I noticed this is on v5 not v7. Is that intentional or did it just get missed?

gabro replied:
just missed, I believe.

isomarcte replied:
Updated

isomarcte commented:
@olafurpg @gabro tests are failing for the deprecated Scala versions on JDK > 8. This is due to the scala compiler's class file parser not having support for parsing some of the constant pool tags, in particular `CONSTANT_MODULE` on these old compiler versions.

What would you like to do? I'd personally be in favor of dropping them. I'm pretty sure metals won't be able to work in general when being used with JDK > 8 class files and Scala < 2.12.10.

isomarcte commented:
> I'd personally be in favor of dropping them.

Dropping them from the cross compilation tests that is.

olafurpg commented:
I'm fine with not testing JDK>8 on Scala <2.12.10

isomarcte commented:
@olafurpg done. This should be good to go now.

isomarcte commented:
@olafurpg the mill tests caused the JVM to core dump. I think they were working on earlier versions of this PR. Can you get those debug logs? I don't think I have access to get them...

olafurpg commented:
@isomarcte I'm unable to access the core dumps

isomarcte commented:
@olafurpg K. I see some non-JDK 11 compliant scala versions being used in the "slow" tests that I missed. Let me fix those and see if that changes things.

olafurpg commented:
FWIW, I recall also hitting on JVM crashes when running the tests locally on JDK 11

isomarcte commented:
@olafurpg I think I've got it all sorted out now.

* As for the JVM core dumps.
    * I think whatever build of `openjdk` that jabba is getting is wonky. I've changed it to `adopt` and things are fine. I noticed the version jabba is grabbing is _very slightly_ out of date, `11.0.2` vs. `11.0.5`.
* The remaining test failures seem to be related to timeouts talking to bloop. I don't think they are related to this PR.

isomarcte commented:
Looks like there are still issues with the LSP integration tests though. I'm looking into it.

isomarcte commented:
@olafurpg @gabro all tests are green.

isomarcte commented:
I rebased the commit and update the commit message. Hopefully all tests stay green.

olafurpg commented on .github/workflows/ci.yml:
```diff
@@ -5,16 +5,36 @@ on:
   pull_request:
 jobs:
+  ubuntuUnitWithJDK8And11:
+    name: ubuntu-latest tests ${{ matrix.java-version }}
```
Instead of duplicating the unit tests for ubuntu, could we try to use `exclude` instead? 🤔 

https://help.github.com/en/actions/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstrategymatrix

```yml
    os: [macos, windows, ubuntu]
    java: [8, 11]
    exclude:
      - os: windows
        java: 8
      - os: macos
        java: 8
```

isomarcte replied:
Sure. Updated.

olafurpg commented on tests/slow/src/test/scala/tests/feature/WorksheetCrossLspSuite.scala:
```diff
@@ -2,5 +2,4 @@ package tests.feature
 
 import scala.meta.internal.metals.{BuildInfo => V}
 
-object Worksheet211LspSuite extends tests.BaseWorksheetLspSuite(V.scala211)
```
Instead of removing these tests for all JVM versions, could we skip them on newer java versions?

isomarcte replied:
Yes. I've updated the code to discriminate on the JDK version and Scala versions.

olafurpg commented on tests/unit/src/main/scala/tests/JVMVersion.scala:
```diff
@@ -0,0 +1,117 @@
+ * modules and will now longer show up as standalone jars,
+ * e.g. `charsets.jar`.
+ */
+sealed trait JVMVersion extends Product with Serializable {
```
Did you try using `scala.util.Properties.isJavaAtLeast`?
```scala
(Scala 2.12.9 Java 1.8.0_222)
@ scala.util.Properties.isJavaAtLeast("9")
res1: Boolean = false


(Scala 2.12.9 Java 11.0.5)
@ scala.util.Properties.isJavaAtLeast("11")
res2: Boolean = true
```

isomarcte replied:
I did not, I did not know about that. I've updated the code, removed the `JVMVersion` class and am now using that predicate.

isomarcte commented:
@olafurpg Updated again. Should be good to go now.

olafurpg commented on tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:
```diff
@@ -5,323 +5,335 @@ import scala.meta.internal.metals.UserConfiguration
-      )
-    } yield ()
-  }
+  scalaVersions.foreach((scalaVersion: String) =>
```
Can we keep the logic simpler here? It’s possible to override “testAsync” and skip tests based on a condition. I’m not a fan of nesting tests this deep

isomarcte replied:
Certainly.

olafurpg commented on tests/slow/src/test/scala/tests/feature/CompletionCrossLspSuite.scala:
```diff
@@ -5,9 +5,6 @@ import tests.BaseCompletionLspSuite
 
 object CompletionCrossLspSuite
     extends BaseCompletionLspSuite("completion-cross") {
-  testAsync("basic-211") {
```
Can we reinstate this test?

olafurpg replied:
It’s fine if not, but I’m curious to know why we can’t keep the 2.11 tests.

isomarcte replied:
@olafurpg I initially removed these because after this PR we will not be running the cross Scala tests on JDK 8. The first version of Scala that supports class file parsing of JDK > 8 bytecode is 2.12.10. Notably, there is no 2.11.x version which supports this AFAIK. 

I can add them back in, and then we can either.

* Discriminate on the JDK version and not run the tests with 2.11.x if we are > 8
* Discriminate on the JDK version and even if it is > 8 only have it emit 8 compatible bytecode.

I'm on the fence about the utility of this myself.

isomarcte replied:
Nevermind, I realized just now that we get the discrimination for free as part of the `BaseCompletionLspSuite` changes. Sorry, I'm still learning my way around the code base.

olafurpg commented on tests/slow/src/test/scala/tests/feature/DefinitionCrossLspSuite.scala:
```diff
@@ -7,10 +7,6 @@ import scala.concurrent.Future
 object DefinitionCrossLspSuite
     extends BaseCompletionLspSuite("definition-cross") {
 
-  testAsync("2.11") {
```
Ditto

olafurpg commented on tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:
```diff
@@ -5,323 +5,335 @@ import scala.meta.internal.metals.UserConfiguration
+      .fold(
+        (e: ScalaVersion.ScalaVersionParseError) => fail(e.errorMessage), {
+          (sv: ScalaVersion) =>
+            if (ScalaVersion.validScalaVersionForCurrentJDK(sv)) {
```
Would be nice to avoid adding a custom “ScalaVersion” class with its own parsing logic just for this condition. Could we solve this problem instead with a hardcoded set of Scala version strings that don’t work in Java 11? 🤔

isomarcte replied:
We can do what you suggest, or we can use more simple parsing logic. It is my opinion both are a bit more brittle and difficult to debug in the case of typos in the input. 

I'll throw up two other branches so we can discuss, but I'll defer to what ever you'd like.

olafurpg commented on tests/unit/src/main/scala/tests/BaseWorksheetLspSuite.scala:
```diff
@@ -5,323 +5,335 @@ import scala.meta.internal.metals.UserConfiguration
 import scala.concurrent.Promise
 
-abstract class BaseWorksheetLspSuite(scalaVersion: String)
+abstract class BaseWorksheetLspSuite(scalaVersions: Set[String])
```
For consistency with the rest of the codebase, I would prefer to keep the old signature and duplicate instead subclasses.

isomarcte replied:
I'll revert that.

isomarcte commented:
@olafurpg I've update the PR to address your requested changes except that I've left `ScalaVersion` in place for now. 

I'll create two more branches to compare to pros/cons of different discrimination strategies on the scala version string. I'm going to be busy for the next 24 hours, so I may not have them up for a day or more.

olafurpg commented on tests/mtest/src/main/scala/tests/ScalaVersion.scala:
```diff
@@ -0,0 +1,99 @@
+ * res0: Boolean = false
+ * }}}
+ */
+final case class ScalaVersion(
```
This class looks like a duplicate of `BuildTool.isCompatibleVersion` https://github.com/scalameta/metals/blob/abca8703638fe2768e0058288ae0e3decbe42e07/metals/src/main/scala/scala/meta/internal/builds/BuildTool.scala#L51

We could add a new package `scala.meta.internal.semver` with a basic helper to compare version numbers. Down the road we can hopefully replace that package with a proper library implementation like Coursier

isomarcte replied:
Do you want me to move `isCompatibleVersion` from `BulidTool` into the `scala.meta.internal.semver` package you propose? Or do you want me to add a dependency on `metals` in the `mtest` project?

olafurpg replied:
Let's move the `isCompatibleVersion` method into a `internal.semver` package inside the `mtags` project so that it can be used from mtest

isomarcte replied:
Updated.

olafurpg commented on tests/mtest/src/main/scala/tests/BaseSuite.scala:
```diff
@@ -37,6 +37,21 @@ class BaseSuite extends TestSuite {
+        (sv: ScalaVersion) => ScalaVersion.validScalaVersionForCurrentJDK(sv)
+      )
+
+  def guardOnScalaVersion(scalaVersion: String)(f: => Unit): Unit =
```
nit: let's remove this helper and use `if (isValidScalaVersion(..)` instead

isomarcte replied:
Once we decide on what to do here (https://github.com/scalameta/metals/pull/1171#discussion_r357985278) I'll remove this.

isomarcte replied:
Updated.

isomarcte commented:
@olafurpg should be ready for review again.

isomarcte commented:
Oops. Looks like I left some unused imports. Should be fixed now. 

Also, FWIW, I removed the `Try` wrapper in the body of `isCompatibleVersion`. Unless I'm _really_ missing something, those inequality checks will never throw.
